File: articles/disability-neurodivergence-and-development/cpp/accessibility_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) values.push_back(item);
    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) return static_cast<int>(i);
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: accessibility_summary data/disability_neurodivergence_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);
    auto header = split_csv_line(line);

    int development_col = find_column(header, "development_score");
    int participation_col = find_column(header, "participation_score");
    int access_col = find_column(header, "current_access");
    int barrier_col = find_column(header, "current_barrier");

    long count = 0;
    double development_sum = 0.0, participation_sum = 0.0, access_sum = 0.0, barrier_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);
        if ((int)values.size() <= development_col || (int)values.size() <= participation_col || (int)values.size() <= access_col || (int)values.size() <= barrier_col) continue;
        development_sum += std::stod(values[development_col]);
        participation_sum += std::stod(values[participation_col]);
        access_sum += std::stod(values[access_col]);
        barrier_sum += std::stod(values[barrier_col]);
        count++;
    }

    std::cout << "Rows analyzed: " << count << "\n";
    std::cout << "Mean development_score: " << development_sum / count << "\n";
    std::cout << "Mean participation_score: " << participation_sum / count << "\n";
    std::cout << "Mean current_access: " << access_sum / count << "\n";
    std::cout << "Mean current_barrier: " << barrier_sum / count << "\n";
    return 0;
}

File: articles/disability-neurodivergence-and-development/go/accessibility_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run accessibility_summary.go data/disability_neurodivergence_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	developmentIdx := col(header, "development_score")
	participationIdx := col(header, "participation_score")
	accessIdx := col(header, "current_access")
	barrierIdx := col(header, "current_barrier")
	if developmentIdx < 0 || participationIdx < 0 || accessIdx < 0 || barrierIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	developmentSum, participationSum, accessSum, barrierSum := 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		development, e1 := strconv.ParseFloat(row[developmentIdx], 64)
		participation, e2 := strconv.ParseFloat(row[participationIdx], 64)
		access, e3 := strconv.ParseFloat(row[accessIdx], 64)
		barrier, e4 := strconv.ParseFloat(row[barrierIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil {
			continue
		}
		count++
		developmentSum += development
		participationSum += participation
		accessSum += access
		barrierSum += barrier
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean development_score: %.4f\n", developmentSum/float64(count))
	fmt.Printf("Mean participation_score: %.4f\n", participationSum/float64(count))
	fmt.Printf("Mean current_access: %.4f\n", accessSum/float64(count))
	fmt.Printf("Mean current_barrier: %.4f\n", barrierSum/float64(count))
}
